docs/shapes.md

```diff
@@ -0,0 +1,193 @@
+`Object.prototype` as `__proto__`. We can see that the shapes inherited the `PropertyTable` from the `root`.
+
+Ok, Let us add a property `x`:
+
+```js
+o.x = 100; // The value is not important!
+```
+
+Then this happens:
+
+```mermaid
+flowchart LR
+    Root(S0: Root Shape\n\nPrototype: None) --->| Property Count: 0 | PropertyTable[PropertyTable\n\nx: Slot 0, writable, configurable, enumerable]
+
+    ObjectPrototype(S1: Prototype Shape\n\nPrototype: Object.prototype) -->| Property Count: 0 |PropertyTable
+    ObjectPrototype -->|Previous| Root
+
+    InsertX(S2: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 1 |PropertyTable
+    InsertX -->|Previous| ObjectPrototype
+
+```
+
+The object `o` has shape `S2` shape now, we can see that it also inherited the `PropertyTable`, but it's property count is `1` and
+an entry has been added into the `PropertyTable`.
+
+We can see that the property added is `writable`, `configurable`, and `enumerable`, but we also see `Slot 0`,
+this is the index into the dense storage in the object itself.
+
+Here is how it would look with the `o` object:
+
+```mermaid
+flowchart LR
+    Root(S0: Root Shape\n\nPrototype: None) --->| Property Count: 0 | PropertyTable[PropertyTable\n\nx: Slot 0, writable, configurable, enumerable]
+
+    ObjectPrototype(S1: Prototype Shape\n\nPrototype: Object.prototype) -->| Property Count: 0 |PropertyTable
+    ObjectPrototype -->|Previous| Root
+
+    InsertX(S2: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 1 | PropertyTable
+    InsertX -->|Previous| ObjectPrototype
+
+    O(Object o\n\nElement 0: JsValue: 100) --> InsertX
+```
+
+Let's define a getter and setter `y`
+
+```js
+// What the getter/setter are not important!
+Object.defineProperty(obj, "y", {
+  enumerable: true,
+  configurable: true,
+  get: function () {
+    return this.x;
+  },
+  set: function (value) {
+    this.x = value;
+  },
+});
+```
+
+```mermaid
+flowchart TD
+    Root(S0: Root Shape\n\nPrototype: None) --->| Property Count: 0 | PropertyTable[PropertyTable\n\nx: Slot 0, writable, configurable, enumerable\ny: Slot 1, has_get, has_set, configurable, enumerable]
+
+    ObjectPrototype(S1: Prototype Shape\n\nPrototype: Object.prototype) -->| Property Count: 0 |PropertyTable
+    ObjectPrototype -->|Previous| Root
+
+    InsertX(S2: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 1 | PropertyTable
+    InsertX -->|Previous| ObjectPrototype
+
+    InsertY(S3: Insert Shape\n\nPrototype: Object.prototype) --> | Property Count: 2 | PropertyTable
+    InsertY -->|Previous| InsertX
+
+    O(Object o\n\nElement 0: JsValue: 100\nElement 1: JsValue: func\nElement 2: JsValue: func) --> InsertY
+```
+
+We can see that the property has been added into the property table, it has the `has_get` and `has_set` flags set,
+in the object there are two elements added, the first is the `get` function and the second is the `set` function.
```

Comment:
Do we want to be careful about using _elements_ for this?
In JSC and V8 elements has the meaning of numeric slots, mainly for arrays, I think in this description we're giving the meaning to any property added? In this case get and set functions. 

Was this inspired from somewhere?

Reply:
I meant it as an element in the storage `vec` in the object itself. Shapes only deal with string and symbol keys, for indexed we have a separate data type ( #2167 ). I can explicitly say `in the storage array in the object` if that is clearer?